Fix/tui aggregated token cost

riturajFi opened the pull request:
**Multi-session token tracking in runtime + sidebar**

---

### **Overview**

fixes #504 

Expands runtime token-usage events to include session IDs and both self vs. inclusive snapshots, allowing downstream consumers to correlate root and child sessions.

The runtime now:

* Aggregates self/child token totals.
* Emits refreshed snapshots after compaction and task transfers.

The sidebar is modernized to:

* Present “Total usage” with comma separators.
* Add a sortable session breakdown.
* Highlight the active agent’s self-spend.

---

### **Implementation Highlights**

* **pkg/runtime/event.go**, **pkg/session/session.go**
  Introduce richer `TokenUsageEvent` payloads and per-session `self` fields, distinguishing the emitting agent’s contribution from inherited child totals.

* **pkg/runtime/runtime.go**
  Calculates self vs. inclusive snapshots, merges child session usage, logs each provider chunk to `token_usage_chunks.log`, and re-emits `TokenUsage` events whenever compaction or transfers change totals.

* **pkg/tui/components/sidebar/sidebar.go**
  Replaces the single `Usage` pointer with a `usageState` map keyed by session ID. Formats totals using `humanize.Comma`, and renders a **“SESSION BREAKDOWN”** section showing root-exclusive usage plus every child session (sorted, with the active session highlighted).

* **pkg/tui/page/chat/chat.go**
  Passes the full `TokenUsageEvent` to the sidebar.

* **pkg/runtime/runtime_test.go**
  Adds snapshot helpers so expectations include the richer event payloads.

---

### **Testing**

```bash
task test
```

---

### **Screenshots / Recordings**

<img width="1110" height="614" alt="image" src="https://github.com/user-attachments/assets/165a2fe9-c439-4633-993d-49920d7c2513" />

riturajFi commented:
@krissetto — kindly take a look at this PR when you have a moment. I’d really appreciate your feedback!

I closed the previous PR since the approach there turned out to be overly complex and didn’t rebase cleanly with the updated main branch.

In this version, I’ve simplified the implementation by introducing dedicated fields for each session to track separate self and child token counts, which are then aggregated for display.

I’ve tested it with:

A root agent having multiple subagents.

Subagents that themselves have nested subagents.

There’s also a temporary logging system included to help with verification — I’ll remove it once the changes are approved for merge.

rumpl commented on pkg/tui/page/chat/chat.go:
```diff
@@ -249,7 +249,7 @@ func (p *chatPage) Update(msg tea.Msg) (layout.Model, tea.Cmd) {
 		return p, cmd
 	case *runtime.TokenUsageEvent:
-		p.sidebar.SetTokenUsage(msg.Usage)
+		p.sidebar.SetTokenUsage(msg) // forward full event so sidebar can track per-session usage
```
Please remove this log

krissetto commented on pkg/tui/components/sidebar/sidebar.go:
```diff
@@ -192,18 +245,49 @@ func (m *model) workingIndicator() string {
-	var usagePercent float64
-	if m.usage.ContextLimit > 0 {
-		usagePercent = (float64(m.usage.ContextLength) / float64(m.usage.ContextLimit)) * 100
+func (m *model) tokenUsageSummary() string { // condensed single-line usage view for horizontal layout
```
follow godoc conventions

krissetto reviewed (commented):
Thanks for working on this @riturajFi! I left a bunch of little comments, mostly nitpicks though.

One thing, i think we could streamline the sidebar UI when no subagents are configured in a team.

Also, have a look [here](https://go.dev/blog/godoc) for best practices when it comes to documenting go code, for the most part we follow those recommendations when leaving comments etc

krissetto commented:
@riturajFi not a problem, i'll leave some comments as i test things out in the meantime. Also, could you squash the commits down when you get a chance?

riturajFi commented:
Oh...i thought you guys wanted to solve it in the backend too as any future consumer would directly have token count access.😅 if just handling things in the frontend is fine then i shall do it right away 🫡

riturajFi commented:
Thanks @krissetto for the greenlight. I shall make the changes in some hours. Kindly find time to take a look.

riturajFi commented:
Hi @krissetto changed the entire logic to the frontend. The changes - 

Added full per-session/-agent usage bookkeeping in pkg/tui/components/sidebar/sidebar.go with session stack tracking, delta cost handling, and global/agent summaries (including horizontal summary string). Token usage updates now take agentName and rely on streamed lifecycle/title events to keep usage aligned with sessions.
Hooked chat page (pkg/tui/page/chat/chat.go) to forward stream start/stop/title and token usage (with agent name) to the sidebar so usage UI stays in sync with streaming.
Adjusted task transfer in pkg/runtime/runtime.go to stop adding child session cost into the parent when delegating, preventing double-counting.
